File: src/main/java/com/kingmang/lazurite/runtime/ClassMethod.java
package com.kingmang.lazurite.runtime;

import com.kingmang.lazurite.parser.AST.Arguments;
import com.kingmang.lazurite.parser.AST.Statements.Statement;
import com.kingmang.lazurite.runtime.values.LzrValue;
import org.jetbrains.annotations.NotNull;

public class ClassMethod extends UserDefinedFunction {
    
    public final ClassInstanceValue classInstance;
    
    public ClassMethod(Arguments arguments, Statement body, ClassInstanceValue classInstance) {
        super(arguments, body);
        this.classInstance = classInstance;
    }
    
    @Override
    public @NotNull LzrValue execute(@NotNull LzrValue... values) {
        Variables.push();
        Variables.define("this", classInstance.getThisMap());
        
        try {
            return super.execute(values);
        } finally {
            Variables.pop();
        }
    }
}

File: src/main/java/com/kingmang/lazurite/runtime/ClassMethod.kt
package com.kingmang.lazurite.runtime

import com.kingmang.lazurite.parser.ast.Arguments
import com.kingmang.lazurite.parser.ast.statements.Statement
import com.kingmang.lazurite.runtime.values.LzrMap
import com.kingmang.lazurite.runtime.values.LzrValue

class ClassMethod(
    arguments: Arguments,
    body: Statement,
    private val lzrClass: ClassInstanceValue
) : UserDefinedFunction(arguments, body) {
    override fun execute(vararg args: LzrValue): LzrValue =
        try {
            Variables.push()
            Variables.define("this", lzrClass)
            super.execute(*args)
        } finally {
            Variables.pop()
        }
}
